packages/Search/src/details/DTK_DetailsTreeTraversal.hpp

```diff
@@ -148,81 +150,131 @@ nearestQuery( BoundingVolumeHierarchy<DeviceType> const &bvh,
+    // immediately popped out of the stack and processed.
+    stack.push( Kokkos::make_pair( root, 0. ) );
 
-    while ( !queue.empty() && count < k )
+    while ( !stack.empty() )
     {
-        // get the node that is on top of the priority list (i.e. is the
-        // closest to the query point)
-        Node const *node = queue.top().first;
-        double const node_distance = queue.top().second;
-        // NOTE: it would be nice to be able to do something like
-        // tie( node, node_distance = queue.top();
-
-        // NOTE: not calling queue.pop() here so that it can be combined with
-        // the next push in case the node is internal (thus sparing a bubble-up
-        // operation)
-        if ( TreeTraversal<DeviceType>::isLeaf( node ) )
-        {
-            queue.pop();
-            insert( TreeTraversal<DeviceType>::getIndex( node ),
-                    node_distance );
-            count++;
-        }
-        else
-        {
-            // insert children of the node in the priority list
+        Node const *node = stack.top().first;
+        double const node_distance = stack.top().second;
+        stack.pop();
 
-            auto const left_child = node->children.first;
-            auto const right_child = node->children.second;
-            queue.pop_push( left_child, distance( left_child ) );
-            queue.push( right_child, distance( right_child ) );
+        if ( node_distance < radius )
+        {
+            if ( TreeTraversal<DeviceType>::isLeaf( node ) )
+            {
+                int const leaf_index =
+                    TreeTraversal<DeviceType>::getIndex( node );
+                double const leaf_distance = node_distance;
+                if ( heap.size() < k )
+                {
+                    // Insert leaf node and update radius if it was the kth one.
+                    heap.push( Kokkos::make_pair( leaf_index, leaf_distance ) );
+                    if ( heap.size() == k )
+                        radius = heap.top().second;
+                }
+                else
+                {
+                    // Replace top element in the heap and update radius.
+                    heap.popPush(
+                        Kokkos::make_pair( leaf_index, leaf_distance ) );
+                    radius = heap.top().second;
+                }
+            }
+            else
+            {
+                // Insert children into the stack and make sure that the
+                // closest one ends on top.
+                Node const *left_child = node->children.first;
+                double const left_child_distance = distance( left_child );
+                Node const *right_child = node->children.second;
+                double const right_child_distance = distance( right_child );
+                if ( left_child_distance < right_child_distance )
+                {
+                    if ( right_child_distance < radius )
```

Comment:
Note that openmp and serial are faster if I remove that condition but then cuda is slower...